modules/ducktests/src/main/java/org/apache/ignite/internal/ducktest/tests/DataGenerationApplication.java

```diff
@@ -18,29 +18,139 @@
+        String cacheName = jsonNode.get(PARAM_CACHE_NAME).asText();
+        boolean infinite = jsonNode.hasNonNull(PARAM_INFINITE) && jsonNode.get(PARAM_INFINITE).asBoolean();
+        boolean optimized = !jsonNode.hasNonNull(PARAM_OPTIMIZED) || jsonNode.get(PARAM_OPTIMIZED).asBoolean();
+        int range = jsonNode.get(PARAM_RANGE).asInt();
+
+        if (infinite) {
+            Random rnd = new Random();
+            CountDownLatch exitLatch = new CountDownLatch(1);
+
+            Thread th = new Thread(() -> {
+                log.info(WATCHEABLE_BEGIN_DATA_GEN_MSG);
+
+                boolean error = false;
+
+                try {
+                    while (!terminated())
+                        generateData(cacheName, range, (idx) -> rnd.nextInt(range), optimized);
+
+                    log.info("Background data generation finished.");
+                }
+                catch (Exception e) {
+                    if (!X.hasCause(e, NodeStoppingException.class)) {
+                        error = true;
+
+                        log.error("Failed to generate data in background.", e);
+                    }
+                }
+                finally {
+                    if (error)
+                        markBroken();
+                    else
+                        markFinished();
+
+                    exitLatch.countDown();
+                }
+
+            }, DataGenerationApplication.class.getName() + "_cacheLoader");
 
-        IgniteCache<Integer, Integer> cache = ignite.createCache(jsonNode.get("cacheName").asText());
+            th.start();
 
-        try (IgniteDataStreamer<Integer, Integer> stmr = ignite.dataStreamer(cache.getName())) {
-            for (int i = 0; i < jsonNode.get("range").asInt(); i++) {
-                stmr.addData(i, i);
+            markInitialized();
 
-                if (i % 10_000 == 0)
-                    log.info("Streamed " + i + " entries");
+            try {
+                exitLatch.await();
             }
+            catch (InterruptedException e) {
+                log.warn("Interrupted waiting for background loading.");
+            }
+        }
+        else {
+            log.info("Generating data...");
+
+            generateData(cacheName, range, Function.identity(), optimized);
+
+            log.info("Data generation finished. Generated " + range + " entries.");
+
+            markSyncExecutionComplete();
         }
+    }
+
+    /** */
+    private void generateData(String cacheName, int range, Function<Integer, Integer> supplier, boolean optimized) {
+        long notifyTime = System.nanoTime();
+        int streamed = 0;
+
+        if(log.isDebugEnabled())
+            log.debug("Creating cache...");
 
-        markSyncExecutionComplete();
+        IgniteCache<Integer, Integer> cache = ignite.getOrCreateCache(cacheName);
+
+        try (IgniteDataStreamer<Integer, Integer> streamer = ignite.dataStreamer(cacheName)) {
+            streamer.allowOverwrite(true);
```

Comment:
Yes, but we don't need this performance for loading. Instead, we need to re-write data sometimes to keep loading.

Reply:
For the most of the tests, we need to load asap.
You may make this an option.

Reply:
> For the most of the tests, we need to load asap.
> You may make this an option.

'Optimized' was involved only for one goal: to work with hard failure of node/data_streamer on kill-signal. An error araises, not exception. It hinders to graceful shutdown, to finish instead of get broken. Current soulution is without 'optimized' option: kust a wrap on the error.